### PR TITLE
Implementation of TH Driver

### PR DESCRIPTION
**Pull request by ghammond86**
TH driver is very similar to Richards driver, and the two could be merged in the future.  To accomplish the TH implementation, it is important that nothing be setup/initialized based on the flow mode until after TDycoreOptions is called as the mode can change at the command line.  I believe that I have segregated all such code.

### CODE REVIEW

**Comment by bishtgautam** [comment]
So, the order of calls should be:
1. `TDySetFromOptions()`
2. `TDySetup()`

- In all demos, there are no calls to any functions between `TDySetFromOptions()` and `TDySetup()`. 
- In `TDydriver.c`, there are calls to functions between  `TDySetFromOptions()` and `TDySetup()` at [L30](https://github.com/TDycores-Project/TDycore/pull/103/files#diff-53e4a62aac64150bcc11cce4e263a97af912847152a02089d9f7ebdc00e29175R30). I moved those intermediate calls before `TDySetFromOptions()`  and the tests ran fine.

Do we expect there could be calls to other functions between `TDySetFromOptions()` and `TDySetup()`?

**Review by bishtgautam — CHANGES_REQUESTED** [review]
I'm getting following warnings:
```
/Users/bish218/projects/tdycore/TDycore/src/tdyth.c: In function 'TDyTHSNESPostCheck':
/Users/bish218/projects/tdycore/TDycore/src/tdyth.c:74:18: warning: unused variable 'ierr' [-Wunused-variable]
   PetscErrorCode ierr;
                  ^~~~
/Users/bish218/projects/tdycore/TDycore/src/tdyth.c: In function 'TDyTHConvergenceTest':
/Users/bish218/projects/tdycore/TDycore/src/tdyth.c:94:7: warning: unused variable 'tdy' [-Wunused-variable]
   TDy tdy = (TDy)ctx;
       ^~~
          CC darwin-gcc8/obj/src/tdyti.o
          CC darwin-gcc8/obj/src/tdydm.o
          CC darwin-gcc8/obj/src/tdyporosity.o
/Users/bish218/projects/tdycore/TDycore/src/tdydriver.c: In function 'TDyDriverInitializeTDy':
/Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:30:3: warning: enumeration value 'TPF' not handled in switch [-Wswitch]
   switch(tdy->method) {
   ^~~~~~
/Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:30:3: warning: enumeration value 'MPFA_O' not handled in switch [-Wswitch]
/Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:67:7: warning: enumeration value 'RICHARDS' not handled in switch [-Wswitch]
       switch (tdy->mode) {
       ^~~~~~
/Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:119:7: warning: enumeration value 'TDySNES' not handled in switch [-Wswitch]
       switch(tdy->ti->time_integration_method) {
       ^~~~~~
/Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:130:7: warning: enumeration value 'TDySNES' not handled in switch [-Wswitch]
       switch(tdy->ti->time_integration_method) {
       ^~~~~~
```

**Comment by ghammond86** [comment]
> So, the order of calls should be:
> 
> 1. `TDySetFromOptions()`
> 2. `TDySetup()`
> 
> * In all demos, there are no calls to any functions between `TDySetFromOptions()` and `TDySetup()`.
> * In `TDydriver.c`, there are calls to functions between  `TDySetFromOptions()` and `TDySetup()` at [L30](https://github.com/TDycores-Project/TDycore/pull/103/files#diff-53e4a62aac64150bcc11cce4e263a97af912847152a02089d9f7ebdc00e29175R30). I moved those intermediate calls before `TDySetFromOptions()`  and the tests ran fine.
> 
> Do we expect there could be calls to other functions between `TDySetFromOptions()` and `TDySetup()`?

There are functions (permeability, porosity, thermal conductivity, etc.) that must be set prior to TDySetup().  These functions are hardwired in the demos.  With the driver, there will be many options for these functions that depend on the command line arguments (TDySetFromOptions).  Thus, they must be set between the calls to TDySetFromOptions and TDySetup.

**Comment by ghammond86** [comment]
> I'm getting following warnings:
> 
> ```
> /Users/bish218/projects/tdycore/TDycore/src/tdyth.c: In function 'TDyTHSNESPostCheck':
> /Users/bish218/projects/tdycore/TDycore/src/tdyth.c:74:18: warning: unused variable 'ierr' [-Wunused-variable]
>    PetscErrorCode ierr;
>                   ^~~~
> /Users/bish218/projects/tdycore/TDycore/src/tdyth.c: In function 'TDyTHConvergenceTest':
> /Users/bish218/projects/tdycore/TDycore/src/tdyth.c:94:7: warning: unused variable 'tdy' [-Wunused-variable]
>    TDy tdy = (TDy)ctx;
>        ^~~
>           CC darwin-gcc8/obj/src/tdyti.o
>           CC darwin-gcc8/obj/src/tdydm.o
>           CC darwin-gcc8/obj/src/tdyporosity.o
> /Users/bish218/projects/tdycore/TDycore/src/tdydriver.c: In function 'TDyDriverInitializeTDy':
> /Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:30:3: warning: enumeration value 'TPF' not handled in switch [-Wswitch]
>    switch(tdy->method) {
>    ^~~~~~
> /Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:30:3: warning: enumeration value 'MPFA_O' not handled in switch [-Wswitch]
> /Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:67:7: warning: enumeration value 'RICHARDS' not handled in switch [-Wswitch]
>        switch (tdy->mode) {
>        ^~~~~~
> /Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:119:7: warning: enumeration value 'TDySNES' not handled in switch [-Wswitch]
>        switch(tdy->ti->time_integration_method) {
>        ^~~~~~
> /Users/bish218/projects/tdycore/TDycore/src/tdydriver.c:130:7: warning: enumeration value 'TDySNES' not handled in switch [-Wswitch]
>        switch(tdy->ti->time_integration_method) {
>        ^~~~~~
> ```

How are you obtaining these warnings?  Which commands line arguments for make?

File: src/tdydriver.c [hunk 1]
```diff
@@ -11,112 +13,140 @@ PetscErrorCode TDyDriverInitializeTDy(TDy tdy) {
...
+//      ierr = TSSetType(ts,TSBEULER); CHKERRQ(ierr);
+//      ierr = TSSetType(ts,TSPSEUDO); CHKERRQ(ierr);
+//      ierr = TSPseudoSetTimeStep(ts,TSPseudoTimeStepDefault,NULL); CHKERRQ(ierr);
+      ierr = TSSetEquationType(ts,TS_EQ_IMPLICIT); CHKERRQ(ierr);
+      ierr = TSSetProblemType(ts,TS_NONLINEAR); CHKERRQ(ierr);
+      ierr = TSSetDM(ts,tdy->dm); CHKERRQ(ierr);
+      ierr = TSSetSolution(ts,tdy->solution); CHKERRQ(ierr);
+      ierr = TDySetIFunction(ts,tdy); CHKERRQ(ierr);
+      ierr = TDySetIJacobian(ts,tdy); CHKERRQ(ierr);
       SNES snes;
```
[the remark this answers is not in the document]
Comment (ghammond86) [reply src/tdydriver.c]: Fixed.

**Review by bishtgautam — APPROVED** [review]

**Pull request merged**